File: problem-2/problem-2.hs
import Data.Function ((&))

iterate2 :: (a -> a -> a) -> a -> a -> [a]
iterate2 f x y = x : iterate2 f y (f x y)

solution :: Integer
solution = fibonacci
    & takeWhile (<= 4000000)
    & filter even
    & sum
  where fibonacci = iterate2 (+) 1 2

main :: IO ()
main = print solution

File: problem-2/problem-2.py
from typing import Generator
from itertools import takewhile

def fibonacci(x: int, y: int) -> Generator[int, None, None]:
    yield(x)
    while True:
        (x, y) = (y, x + y)
        yield(x)

print(sum(
    filter(
        lambda n: n % 2 == 0,
        takewhile(
            lambda n: n <= 4_000_000,
            fibonacci(1, 2)
        )
    )
))
